frontend/src/metabase-lib/lib/queries/StructuredQuery.ts

```diff
@@ -303,12 +303,21 @@ export default class StructuredQuery extends AtomicQuery {
     const sourceQuery = this.sourceQuery();
 
     if (sourceQuery) {
+      const dimensions = sourceQuery.dimensions();
+      const dimensionFields = dimensions.map(dimension => dimension.field());
       return new Table({
         name: "",
         display_name: "",
         db: sourceQuery.database(),
-        fields: sourceQuery.columns().map(
-          column =>
+        fields: sourceQuery.columns().map(column => {
+          // FIX for https://github.com/metabase/metabase/issues/12985
+          // retrieve dimension and generate field if it is matching the column
+          const dimensionField = dimensionFields.find(({ name, base_type }) => {
```

Comment:
Retrieving the field (from the dimensions) based on the column: `name` and `base_type`.